docs(styling): provide theming documentation (#394)
IgniteUI/igniteui-xplat-docs: doc/en/components/themes/palettes.md

Here's an excerpt of the `primary` color as declared in the Light Bootstrap Palette:

```css
:root {
  --igc-primary-h: 216deg;
  --igc-primary-s: 98%;
  --igc-primary-l: 52%;
  --igc-primary-a: 1;
  --igc-primary-50: hsla(
                      var(--igc-primary-h),
                      calc(var(--igc-primary-s) * 1.23),
                      calc(var(--igc-primary-l) * 1.78),
                      var(--igc-primary-a)
                    );
  --igc-primary-100: hsla(
                      var(--igc-primary-h),
                      calc(var(--igc-primary-s) / 1.25),
                      calc(var(--igc-primary-l) * 1.66),
                      var(--igc-primary-a)
                    );
  --igc-primary-200: hsla(
                      var(--igc-primary-h),
                      calc(var(--igc-primary-s) * .64),
                      calc(var(--igc-primary-l) * 1.43),
                      var(--igc-primary-a)
                    );
  // ...
  --igc-primary-50-contrast: black;
  --igc-primary-100-contrast: black;
  --igc-primary-200-contrast: black;
  // ...
}
```

> All color variants are derived from 4 base variables - `--igc-primary-h`, `--igc-primary-s`, `--igc-primary-l`, and `--igc-primary-a`. Each one of these variables holds the [HSLA](https://drafts.csswg.org/css-color/#the-hsl-notation) parts for a single color. HSLA stands for `hue`, `saturation`, `lightness`, and `alpha`. It's another color scheme used to describe colors. We decided to use this approach as it allows us to modify all variants of the `primary`, `secondary` and other colors at runtime.

Reviewer: ```suggestion
All color variants are derived from 4 base variables - `--igc-primary-h`, `--igc-primary-s`, `--igc-primary-l`, and `--igc-primary-a`. Each one of these variables holds the [HSLA](https://drafts.csswg.org/css-color/#the-hsl-notation) parts for a single color. HSLA stands for `hue`, `saturation`, `lightness`, and `alpha`. It's another color space used to represent colors. We decided to use this approach as it allows us to modify all variants of the `primary`, `secondary` and other colors at runtime.
```

## Defining Palettes

If you wanted to change the color variants for a color from the palette, you can do so by overriding its HSLA values in your stylesheet. For instance, changing the primary colors is as easy as writing:

```css
/* The HSLA representation of orange (#ffa500) */
/* hsla(38.8, 100%, 50%, 1); */
:root {
  --igc-primary-h: 38.8deg; 
  --igc-primary-s: 100%; 
  --igc-primary-l: 50%; 
  --igc-primary-a: 1; 
}
```

This will automatically update all the other primary variants.

You will notice that color variants for each color are monochromatic. This is because all color variants are generated from the HSLA variables. You can override individual colors only using any color scheme:

> ```css
> :root {
>   --igc-primary-600: darkorange;
> }
> ```

Reviewer: This has to remain in the HSLA color space, unfortunately. Further, it has to be in the format of list of values for h, s, l, and a.

```suggestion
```css
:root {
  --igc-primary-600: 32.9deg ,100%, 50%, 1;
}
```
```

Be cautious when doing this:

```css
:root {
  --igc-primary-500: orange;
  --igc-primary-600: blue;
}

> ```

Reviewer: This needs to be updated to match the example above.

It may result in unexpected results, as some component themes use more than one color variant. We designed all component themes around monochromatic palettes.

## Scoping

We've seen that overriding colors in the palette is relatively easy. We can update the _global_ palette by scoping color variants to the `:root` selector in the `styles.css` file of our application.

Let's say your corporate primary color is `#9f349c` and you want to create primary variants for it. One option would be to use the [Material Color Tool](https://material.io/design/color/the-color-system.html#tools-for-picking-colors) to generate all color variants for you. Here's how we will declare the produced colors by the tool:

```css
:root {
  --igc-primary-50: #f3e6f2;
  --igc-primary-100: #e2c0e0;
  --igc-primary-200: #d096cd;
  --igc-primary-300: #d096cd;
  --igc-primary-400: #bd6db9;
  --igc-primary-500: #9f349c;
  --igc-primary-600: #923095;
  --igc-primary-700: #81298d;
  --igc-primary-800: #722584;

>   --igc-primary-900: #561d74;

Reviewer: Again, this has to follow the format mentioned above.

}
```

> This approach disregards the `calc` function we use for coming up with color variants from HSLA scheme at runtime, however, it allows you to specify a hand-picked palette.

Reviewer: ```suggestion
```

```css
/* cornflowerblue hsl(218.5, 79.2%, 66.1%) */
.blue-theme {
  --igc-primary-h: 218.5deg; 
  --igc-primary-s: 79.2%; 
  --igc-primary-l: 66.1%; 
  --igc-primary-a: 1; 
}

/* brick red hsl(351.7, 57%, 52.5%) */
.red-theme {
  --igc-primary-h: 351.7deg; 
  --igc-primary-s: 57%; 
  --igc-primary-l: 52.5%; 
  --igc-primary-a: 1; 
}
```

Then you can simply overhaul the colors in your application by changing the value of the class attribute from `blue-theme` to `red-theme`.

## Dark vs. Light

Palettes in $ProductName$ dictate whether a theme is going to be light or dark. The two colors that have the biggest impact on that are `gray` and `surface`. See, the `gray` color variants in all themes are based on either a very light color shade, like `#fff`, or a very dark one like `#222`. Light themes have `gray` variants based on dark shades of gray, while dark themes are the opposite - all `gray` variants are a shade of white. These `gray` colors will be displayed against another color, usually the `surface` color. The `surface` color should always be on the opposite end of the `gray` in the gray scale to ensure themes look good.

> To make this a bit clearer, below is the complete list of all `gray` and `surface` color variants in both a light and a dark theme as represented in the RGBA color space.

Reviewer: ```suggestion
To make this a bit clearer, below is the complete list of all `gray` and `surface` color variants in both a light and a dark theme.
```

*Material Light:*
```css
:root {
  /* surface is set to white */
  --igc-surface-500: #fff;

  /* gray are based on black to contrast the surface color */
  --igc-gray-50: rgba(0, 0, 0, 0.02);
  --igc-gray-100: rgba(0, 0, 0, 0.04);
  --igc-gray-200: rgba(0, 0, 0, 0.08);
  --igc-gray-300: rgba(0, 0, 0, 0.12);
  --igc-gray-400: rgba(0, 0, 0, 0.26);
  --igc-gray-500: rgba(0, 0, 0, 0.38);
  --igc-gray-600: rgba(0, 0, 0, 0.54);
  --igc-gray-700: rgba(0, 0, 0, 0.62);
  --igc-gray-800: rgba(0, 0, 0, 0.74);

>   --igc-gray-900: rgba(0, 0, 0, 0.87);

Reviewer: Same as above.

}
```

*Material Dark:*
```css
:root {
  /* surface is a dark shade of gray */
  --igc-surface-500: #222;

  /* gray are based on white to contrast the surface color */
  --igc-gray-50: rgba(255, 255, 255, 0.02);
  --igc-gray-100: rgba(255, 255, 255, 0.04);
  --igc-gray-200: rgba(255, 255, 255, 0.08);
  --igc-gray-300: rgba(255, 255, 255, 0.12);
  --igc-gray-400: rgba(255, 255, 255, 0.26);
  --igc-gray-500: rgba(255, 255, 255, 0.38);
  --igc-gray-600: rgba(255, 255, 255, 0.54);
  --igc-gray-700: rgba(255, 255, 255, 0.62);
  --igc-gray-800: rgba(255, 255, 255, 0.74);

>   --igc-gray-900: rgba(255, 255, 255, 0.87);

Reviewer: Same as above.

}
```

Be mindful when changing the `gray` and `surface` color variants as they are used in most components and have a big impact on their overall look and feel. 

## Other Colors

So far we've covered the `primary`, `secondary`, `gray`, and `surface` color variants and how you can override them. There are four more colors - `info`, `success`, `warn`, and `error`. They are usually used to set the colors in different states. For example, the `igc-input-group` component uses these colors in its input validation states.

```css
:root {
  --igc-info-500: blue;
  --igc-success-500: green;
  --igc-warn-500: orange;

>   --igc-error-500: red;

Reviewer: Same as above.
